README.md
# Kong database importer

## Summary

This tool was created to help speed up load testing Kong deployments. When
load testing very large deployments it can take a while to create all of the
entities in via Deck or the Kong admin API. This tool will directly create
Kong entities in the Kong database. It only supports Kong deployments 
with a postgresql database.

### Disclaimer

This is tool is a prototype and is not supported by anyone. Use at your own
risk. Bypassing the Kong admin APIs and directly importing data into the
Kong database is a fragile process and is prone to breaking between releases,
and causing unpredictable behaviour.

## Installation

1. clone this repo or extract the zip containing the repo
2. install required modules using pip (see requirements.txt) 
   `pip install -r requirements.txt`
3. create a config file (YAML format) that describes how many entities to
   create in the Kong database. See the Config File section for details
3. run the application (see the Running the Script section)

## Running the Script

Create example:
```
  python ./runner.py \
	--config-file ./config.yaml \
	--hostname 127.0.0.1 \
	--database kong \
	--username kong \
  --password p@55word
```

When run with the above options, the tooling will read and process the config
file (config.yaml). Then it will connect to the database and start creating
entities directly via the [psycopg](https://www.psycopg.org/) library.
The tooling makes use of this of the bulk `COPY` import method to make the
process as efficient as possible.

Delete example:
```
  python ./runner.py \
	--config-file ./config.yaml \
	--hostname 127.0.0.1 \
	--database kong \
	--username kong \
  --password p@55word \
  --delete
```

The `--delete` flag can be passed on the command line and the tooling will
then remove any entities described in the config file from the Kong database


## Config File

The config file should be in YAML format, an example of the config file with
all options specifed is shown below.

```yaml
workspaces: 100
prefix: perf
consumers_per_workspace: 50
services_per_workspace: 50
service_protocol: http
service_host: httpbin.org
service_port: 80
service_path: /
routes_per_service: 7
plugins:
  file-log:
    config:
      path: "/dev/null"
      custom_fields_by_lua: {}
      reopen: false
  cors:
    config:
      max_age: null
      credentials: false
      exposed_headers: null
      methods: [ "GET" ]
      headers: null
      preflight_continue: false
      origins: [ "*" ]
  ip-restriction:
    config:
      message: null
      allow: null
      deny: [ "1.1.1.1" ]
      status: null
  udp-log:
    config:
      timeout: 0
      port: 9999
      host: "127.0.0.1"
      custom_fields_by_lua: {}

```

As you can see the number of workspaces is dictated by the `workspaces` field.
The workspaces are named with the `prefix-<number>` notaition. so with 100
workspaces specifed and a prefix of perf, you will end up with workspaces 
named `perf-workspace-0` to `perf-workspace-99`. `consumers`, `services` and
`routes` are all declared using a number. Setting `services_per_workspace`
to 50 will create 50 services per workspace, with the name of the service formed
from the workspace and the iteration number of the particular service
e.g. `perf-workspace-37-svc-15`. Routes are created per service do a `routes_per_service`
setting of 7 will create 7 routes for every service in every workspace.
Consumers are created on a workspace level, setting `consumers_per_workspace` to
50 will create 50 consumer in every workspace created by this tool. Consumers are
named as follows: `perf-workspace-39-consumer-43` for the 43rd consumer in the 39th
workspace. Lastly there is plugins, Plugins differ in that there configuration
properties differ wildly, and also teams want flexibility when load testing with
regrads to the plugins they want to load test with. With this in mind the plugins
config is a hash, with the plugin name as the key and then a config key is nested
under the plugin name key. Under the config key you should put all of the configuration
for that particular plugin - even the default or null values. The plugins in the hash
will be applied to every service.


CHANGELOG.md
# Changelog

## 0.0.1

🆕 New features:
- Project created

- CHANGELOG added

([PR #1](https://github.com/srb3/kong-postgres-importer/pull/1))


tests/fixtures/integration/start.sh
#!/usr/bin/env bash

topo="${1:-kong_gateway}"
control="${DMODE:-up}"
kong_version="${KONG_VERSION:-2.8.1.1-alpine}"
kong_plugins="${PLUGINS:-bundled}"
kong_config="${DECK_CONFIG}"
admin_url="${ADMIN_URL:-http://localhost:8001}"
token="${TOKEN:-password}"
workspace="${WORKSPACE:-default}"
script="${SCRIPT}"
quiet="${QUIET}"

base_path=$(cd -- "$(dirname -- "${BASH_SOURCE[0]}")" &>/dev/null && pwd)
compose_dir="./compose_topologies"

remove_splunk_file() {
	pushd "${base_path}" || exit
	if [[ -f .splunk_token ]]; then
		echo "removing splunk token"
		rm -f .splunk_token
	fi
}

wait_connect() {
	curl --retry-all-errors --connect-timeout 5 \
		--max-time 10 \
		--retry 5 \
		--retry-delay 0 \
		--retry-max-time 40 \
		"$admin_url" &>>/dev/null
}

deck_sync() {
	wait_connect
	if [[ -d "${base_path}/dumps/$kong_config/workspaces" ]]; then
		pushd "${base_path}/dumps/$kong_config/workspaces" || exit
		for i in $(ls); do
			if [[ -n "${quiet}" ]]; then
				deck sync \
					--kong-addr "$admin_url" \
					--headers "kong-admin-token: $token" \
					-s "$i" &>/dev/null
			else
				deck sync \
					--kong-addr "$admin_url" \
					--headers "kong-admin-token: $token" \
					-s "$i"
			fi
		done
		popd || exit
	else
		deck sync \
			--kong-addr "$admin_url" \
			--headers "kong-admin-token: $token" \
			--workspace "$workspace" \
			-s "${base_path}/dumps/$kong_config/split"
	fi
}

post_script() {
	wait_connect
	pushd "${base_path}" || exit
	if [[ -n "${quiet}" ]]; then
		bash "${base_path}/scripts/${script}/run.sh" &>/dev/null
	else
		bash "${base_path}/scripts/${script}/run.sh"
	fi
	popd || exit
}

up() {
	compose_up
}

down() {
	compose_down
}

compose_up() {
	docker-compose -f docker-compose.yml \
		-f postgres/postgres.yml \
		-f kong-bootstrap/kong-bootstrap.yml \
		-f kong-control-plane/kong-control-plane.yml \
		-f kong-data-plane/kong-data-plane.yml \
		-f pgadmin/pgadmin.yml \
		up -d
}

compose_down() {
	docker-compose -f docker-compose.yml \
		-f postgres/postgres.yml \
		-f kong-bootstrap/kong-bootstrap.yml \
		-f kong-control-plane/kong-control-plane.yml \
		-f kong-data-plane/kong-data-plane.yml \
		-f pgadmin/pgadmin.yml \
		down
}

run() {
	if [[ "$control" == "up" ]]; then
		up
		if [[ -n "$kong_config" ]] && [[ "$topo" != "dbless" ]]; then
			echo "skipping deck_sync"
		fi
		if [[ -n "$script" ]]; then
			post_script
		fi
	elif [[ "$control" == "down" ]]; then
		echo "calling down"
		down
		remove_splunk_file
	elif [[ "$control" == "restart-dbless" ]]; then
		docker-compose -f docker-compose.yml \
			-f kong-dbless/kong-dbless.yml \
			restart kong-dbless
	elif [[ "$control" == "restart-dp-1" ]]; then
		docker-compose restart kong-data-plane-1
	elif [[ "$control" == "restart-dp-2" ]]; then
		docker-compose restart kong-data-plane-2
	elif [[ "$control" == "restart-cp" ]]; then
		docker-compose restart kong-control-plane
	elif [[ "$control" == "restart-kong" ]]; then
		docker-compose restart kong
	elif [[ "$control" == "restart" ]]; then
		echo "restarting"
		down
		up
		if [[ -n "$kong_config" ]] && [[ "$topo" != "dbless" ]]; then
			deck_sync
		fi
		if [[ -n "$script" ]]; then
			post_script
		fi
	fi
}

kong_gateway() {
	local path="${compose_dir}/${topo}"
	export KONG_PLUGINS="$kong_plugins"
	export KONG_VERSION="$kong_version"
	pushd "$path" || exit
	echo "calling run from ${topo}"
	run
	popd || exit
}

echo "CI_MODE IS: ${CI_MODE}"

$topo


tests/fixtures/integration/topology.sh
#!/usr/bin/env bash
MODE=${1:-up}
STATE=${2:-identity-bridge-demo}
TOPOLOGY=${3:-kong_gateway}
IMAGE_BUILD=${4:-local}
TPATH=${TERRAFORM_PATH:-docker-with-azure-ad}
IMAGE_VERSION=${IMAGE_VERSION:-2.8.1.1}

export DECK_LOG_TIMEOUT=${DECK_LOG_TIMEOUT:-10000}
export DECK_LOG_RETRY_COUNT=${DECK_LOG_RETRY_COUNT:-3}
export DECK_LOG_QUEUE_SIZE=${DECK_LOG_QUEUE_SIZE:-4000}

# echo "log timeout set to: $DECK_LOG_TIMEOUT"

if [[ "${IMAGE_BUILD}" != "local" ]]; then
	image_build="${IMAGE_BUILD}"
else
	image_build="local"
fi

if [[ "${MODE}" == "up" ]]; then
	if [[ -z "${SKIP_TF}" ]]; then
		pushd "./terraform/${TPATH}" || exit
		terraform init
		terraform apply -auto-approve
		popd || exit
	fi
	source "./terraform/${TPATH}/temp_files/idp_creds_env"
	source "./terraform/${TPATH}/temp_files/env_auth_conf"
elif [[ "${MODE}" == "down" ]]; then
	if [[ -z "${SKIP_TF}" ]]; then
		pushd "./terraform/${TPATH}" || exit
		terraform destroy --auto-approve
		popd || exit
	fi
fi

if [[ -n "${QUIET}" ]]; then
	export QUIET=true
fi

export KONG_PORTAL_APP_AUTH=kong-oauth2

if [[ "${image_build}" == "local" ]]; then
	IMAGE=kong/kong-gateway
	VERSION=${IMAGE_VERSION}-rhel7
elif [[ "${image_build}" == "test" ]]; then
	IMAGE=kongcx/kong-test-custom
	VERSION="${IMAGE_VERSION}-rhel7-${TEST_VERSION}"
else
	IMAGE=kongcx/kong-custom
	VERSION="${IMAGE_VERSION}-rhel7-${image_build}"
fi

if [[ "${TOPOLOGY}" == "kong_gateway" ]]; then
	QUIET="${QUIET}" \
		ACCESS_LOG="/dev/stdout" \
		ERROR_LOG="/dev/stderr" \
		SCRIPT=${STATE} \
		DECK_CONFIG=${STATE} \
		KONG_LOG_LEVEL=debug \
		KONG_IMAGE=${IMAGE} \
		KONG_VERSION=${VERSION} \
		PLUGINS=bundled,splunk-hec,rate-limiting-advanced-sidecar,gateway-identity-bridge,sidecar-identity-bridge,request-termination-ext \
		DMODE=${MODE} \
		./start.sh "${TOPOLOGY}"
fi


tests/fixtures/plugins/splunk-hec/schema.lua
-- This software is copyright Kong Inc. and its licensors.
-- Use of the software is subject to the agreement between your organization
-- and Kong Inc. If there is no such agreement, use is governed by and
-- subject to the terms of the Kong Master Software License Agreement found
-- at https://konghq.com/enterprisesoftwarelicense/.
-- [ END OF LICENSE 0867164ffc95e54f04670b5169c09574bdbd9bba ]

local typedefs = require "kong.db.schema.typedefs"
local url = require "socket.url"

return {
  name = "http-log",
  fields = {
    { protocols = typedefs.protocols },
    { config = {
        type = "record",
        fields = {
          -- NOTE: any field added here must be also included in the handler's get_queue_id method
          { http_endpoint = typedefs.url({ required = true }) },
          { method = { type = "string", default = "POST", one_of = { "POST", "PUT", "PATCH" }, }, },
          { content_type = { type = "string", default = "application/json", one_of = { "application/json" }, }, },
          { token = { type = "string" }, },
          { debug = { type = "boolean", required = true, default = false }, },
          { timeout = { type = "number", default = 10000 }, },
          { keepalive = { type = "number", default = 60000 }, },
          { retry_count = { type = "integer", default = 10 }, },
          { queue_size = { type = "integer", default = 1 }, },
          { flush_timeout = { type = "number", default = 2 }, },
          { headers = typedefs.headers {
            keys = typedefs.header_name {
              match_none = {
                {
                  pattern = "^[Hh][Oo][Ss][Tt]$",
                  err = "cannot contain 'Host' header",
                },
                {
                  pattern = "^[Cc][Oo][Nn][Tt][Ee][Nn][Tt]%-[Ll][Ee][nn][Gg][Tt][Hh]$",
                  err = "cannot contain 'Content-Length' header",
                },
                {
                  pattern = "^[Cc][Oo][Nn][Tt][Ee][Nn][Tt]%-[Tt][Yy][Pp][Ee]$",
                  err = "cannot contain 'Content-Type' header",
                },
              },
            },
          } },
          { custom_fields_by_lua = typedefs.lua_code },
        },
        custom_validator = function(config)
          -- check no double userinfo + authorization header
          local parsed_url = url.parse(config.http_endpoint)
          if parsed_url.userinfo and config.headers then
            for hname, hvalue in pairs(config.headers) do
              if hname:lower() == "authorization" then
                return false, "specifying both an 'Authorization' header and user info in 'http_endpoint' is not allowed"
              end
            end
          end
          return true
        end,
      },
    },
  },
}


runner.py
import psycopg
import json
import uuid
import yaml
import re
from typing import Dict, List, Any


class Runner(object):
    """
    A class used to orchestrate the parsing of config into kong entites,
    and then pass those entities into kong

    Attributes
    ----------
    config_file : string
        a path to a config file containing the types and ammount of
        entity to create
    db_params: dict
        A dictionary of database connection parameters, used for accessing
        the Kong database

    Methods
    -------
    """

    def __init__(
        self,
        config_file=None,
        db_params=None,
        delete=False,
        route_dump=False,
        route_dump_location=None,
    ) -> None:
        """
        Parameters
        ----------
        config_file : str
            The Path to the modules config file (YAML)
        db_params: dict
            A dictionary of database connection parameters, used for accessing
            the Kong database
        """
        self.config_file: str = config_file
        self.db_params: Dict[str, str] = db_params
        self.data: Dict[str, Any] = self.parse_config(self.config_file)
        self.prefix: str = self.data["prefix"]
        self.delete: bool = delete
        self.route_dump: bool = route_dump
        self.route_dump_location = route_dump_location or "./routes.josn"
        self.required_workspaces: List[List[str]] = self.gen_workspaces(
            self.data["workspaces"]
        )
        self.required_workspace_names: List[str] = [
            v[1] for v in self.required_workspaces
        ]

        self.number_of_services: int = int(self.data["services_per_workspace"])
        self.number_of_routes: int = int(self.data["routes_per_service"])
        self.number_of_consumers: int = int(self.data["consumers_per_workspace"])
        self.number_of_plugins: int = len(self.data["plugins"]) or 0
        self.plugins = self.data["plugins"] or {}
        self.svc_defaults = self.get_svc_defaults(self.data)

        self.entites: List[str] = [
            "workspaces",
            "services",
            "routes",
            "consumers",
            "plugins",
        ]

        self.get_items: Dict[str, List[str]] = {
            "workspaces": ["name", "id"],
            "services": ["id", "name", "ws_id"],
            "routes": ["id", "name", "service_id", "ws_id"],
            "plugins": ["id", "name", "service_id", "cache_key", "ws_id"],
            "consumers": ["id", "username", "ws_id"],
        }

        self.insert_items: Dict[str, List[str]] = {
            "workspaces": ["id", "name", "meta"],
            "services": [
                "id",
                "name",
                "retries",
                "protocol",
                "host",
                "port",
                "path",
                "connect_timeout",
                "write_timeout",
                "read_timeout",
                "ws_id",
                "enabled",
            ],
            "routes": [
                "id",
                "name",
                "service_id",
                "protocols",
                "paths",
                "regex_priority",
                "strip_path",
                "preserve_host",
                "https_redirect_status_code",
                "path_handling",
                "ws_id",
                "request_buffering",
                "response_buffering",
            ],
            "plugins": [
                "id",
                "name",
                "service_id",
                "config",
                "enabled",
                "cache_key",
                "protocols",
                "ws_id",
            ],
            "consumers": [
                "id",
                "username",
                "ws_id",
                "username_lower",
                "type",
            ],
        }

        self.get_entity_data: Dict[str, Any] = {
            "workspaces": self.workspaces_parser,
            "services": self.services_parser,
            "routes": self.routes_parser,
            "consumers": self.consumers_parser,
            "plugins": self.plugins_parser,
        }

        self.entity_data_hydrate: Dict[str, Any] = {
            "workspaces": self.workspaces_data_hydrate,
            "services": self.services_data_hydrate,
            "routes": self.routes_data_hydrate,
            "consumers": self.consumers_data_hydrate,
            "plugins": self.plugins_data_hydrate,
        }

        self.db_cache: Dict[str, Any] = {
            "workspaces": {},
            "services": {},
            "routes": {},
            "consumers": {},
            "plugins": {},
        }

        self.parse_cache: Dict[str, Any] = {
            "workspaces": {},
            "services": {},
            "routes": {},
            "consumers": {},
            "plugins": {},
        }

        self.composit_cache: Dict[str, Any] = {
            "workspaces": [],
            "services": [],
            "routes": [],
            "consumers": [],
            "plugins": [],
        }

        self.require_strings: Dict[str, Any] = {
            "workspaces": self.workspaces_require_string,
            "services": self.services_require_string,
            "routes": self.routes_require_string,
            "consumers": self.consumers_require_string,
            "plugins": self.plugins_require_string,
        }
        if self.route_dump:
            print("dumping routes")
            with open(self.route_dump_location, "w") as f:
                json.dump(self.dump_routes(), f)
        if self.delete:
            print("deleting entities")
            self.delete_entities()
        else:
            print("creating entities")
            self.create_entities()

    ################# Data parsers #############################

    def workspaces_require_string(self) -> List[str]:
        return ["{}^{}".format(v[0], v[1]) for v in self.required_workspaces]

    def consumers_require_string(self) -> List[str]:
        return [
            "{}-consumer-{}^{}".format(k, i, v["id"])
            for i in range(self.number_of_consumers)
            for k, v in self.get_active_entities_data("workspaces").items()
        ]

    def services_require_string(self) -> List[str]:
        return [
            "{}-svc-{}^{}".format(k, i, v["id"])
            for i in range(self.number_of_services)
            for k, v in self.get_active_entities_data("workspaces").items()
        ]

    def dump_routes(self) -> List[str]:
        routes = []
        for ws in self.required_workspace_names:
            for s in range(self.number_of_services):
                for r in range(self.number_of_routes):
                    routes.append("{}-svc-{}-route-{}".format(ws, s, r))
        return routes

    def routes_require_string(self) -> List[str]:
        return [
            "{}-route-{}^{}^{}".format(v["name"], i, k, v["ws_id"])
            for i in range(self.number_of_routes)
            for k, v in self.get_active_entities_data("services").items()
        ]

    def plugins_require_string(self) -> List[str]:
        return [
            "{}^{}^{}".format(p, k, v["ws_id"])
            for p in self.plugins
            for k, v in self.get_active_entities_data("services").items()
        ]

    def workspaces_data_hydrate(self, comp: str = ""):
        p = comp.split("^")
        id_ = p[0]
        name = p[1]
        meta = '{"color": "#3894f0"%"thumbnail": null}'
        return "{},{},{}".format(id_, name, meta)

    def services_data_hydrate(self, comp: str):
        p = comp.split("^")
        name = p[0]
        ws_id = p[1]
        retries = self.svc_defaults["retries"]
        protocol = self.svc_defaults["protocol"]
        host = self.svc_defaults["host"]
        port = self.svc_defaults["port"]
        path = self.svc_defaults["path"]
        connect_timeout = self.svc_defaults["connect_timeout"]
        write_timeout = self.svc_defaults["write_timeout"]
        read_timeout = self.svc_defaults["read_timeout"]
        return "{},{},{},{},{},{},{},{},{},{},{},true".format(
            uuid.uuid4(),
            name,
            retries,
            protocol,
            host,
            port,
            path,
            connect_timeout,
            write_timeout,
            read_timeout,
            ws_id,
        )

    def consumers_data_hydrate(self, comp: str):
        p = comp.split("^")
        username = p[0]
        ws_id = p[1]
        username_lower = username
        type_ = "0"
        return "{},{},{},{},{}".format(
            uuid.uuid4(), username, ws_id, username_lower, type_
        )

    def routes_data_hydrate(self, comp: str):
        p = comp.split("^")
        name = p[0]
        service_id = p[1]
        ws_id = p[2]
        return (
            "{},{},{},{{http% https}},{{/{}}},0,true,false,426,v0,{},true,true".format(
                uuid.uuid4(),
                name,
                service_id,
                name,
                ws_id,
            )
        )

    def plugins_data_hydrate(self, comp: str):
        p = comp.split("^")
        name = p[0]
        service_id = p[1]
        ws_id = p[2]
        cache_key = "plugins:{}::{}:::{}".format(name, service_id, ws_id)
        return "{},{},{},CONFIG,true,{},{{grpc%grpcs%http%https}},{}".format(
            uuid.uuid4(),
            name,
            service_id,
            cache_key,
            ws_id,
        )

    def workspaces_parser(
        self,
        current: Dict[str, Any],
        required: List[str],
        entity: str,
    ) -> Dict[str, Any]:
        parsed_data = self.parse_cache[entity]

        if parsed_data:
            return parsed_data
        else:
            self.parse_cache[entity] = {
                k: v for k, v in current.items() if v["name"] in required
            }
            return self.parse_cache[entity]

    def services_parser(
        self, current: Dict[str, Any], active: List[str], entity: str
    ) -> Dict[str, Any]:
        parsed_data = self.parse_cache[entity]
        if parsed_data:
            return parsed_data
        else:
            return {
                v["id"]: {
                    "name": v["name"],
                    "ws_id": v["ws_id"],
                    "composit": "{}^{}".format(v["name"], v["ws_id"]),
                }
                for v in current.values()
                if v["ws_id"] in active
            }

    def routes_parser(
        self, current: Dict[str, Any], active: List[str], entity: str
    ) -> Dict[str, Any]:
        parsed_data = self.parse_cache[entity]
        if parsed_data:
            return parsed_data
        else:
            return {
                v["id"]: {
                    "name": v["name"],
                    "ws_id": v["ws_id"],
                    "composit": "{}^{}^{}".format(
                        v["name"], v["service_id"], v["ws_id"]
                    ),
                }
                for v in current.values()
                if v["ws_id"] in active
            }

    def consumers_parser(
        self, current: Dict[str, Any], active: List[str], entity: str
    ) -> Dict[str, Any]:
        parsed_data = self.parse_cache[entity]
        if parsed_data:
            return parsed_data
        else:
            return {
                v["id"]: {
                    "username": v["username"],
                    "composit": "{}^{}".format(v["username"], v["ws_id"]),
                }
                for v in current.values()
                if v["ws_id"] in active
            }

    def plugins_parser(
        self, current: Dict[str, Any], active: List[str], entity: str
    ) -> Dict[str, Any]:
        parsed_data = self.parse_cache[entity]
        if parsed_data:
            return parsed_data
        else:
            return {
                v["id"]: {
                    "name": v["name"],
                    "composit": "{}^{}^{}".format(
                        v["name"], v["service_id"], v["ws_id"]
                    ),
                }
                for v in current.values()
                if v["ws_id"] in active
            }

    ################### Helper functions #########################

    def parse_config(self, config_file) -> Dict[str, Any]:
        d = {}
        with open(config_file, "r") as stream:
            try:
                d = yaml.safe_load(stream)
            except yaml.YAMLError as exc:
                print(exc)
        return d

    def get_svc_defaults(self, data: Dict[str, str]) -> Dict[str, str]:
        return {
            "protocol": self.set_param("service_protocol", "http", data),
            "host": self.set_param("service_host", "httpbin.org", data),
            "port": self.set_param("service_port", "80", data),
            "path": self.set_param("service_path", "/", data),
            "retries": self.set_param("service_retries", "5", data),
            "connect_timeout": self.set_param("service_connect_timeout", "60000", data),
            "write_timeout": self.set_param("service_write_timeout", "60000", data),
            "read_timeout": self.set_param("service_read_timeout", "60000", data),
        }

    def set_param(self, value: str, default: str, data: Dict[str, str]) -> str:
        return data[value] if value in data.keys() else default

    def db_connect(self, params: Dict[str, str]) -> str:
        hostname = self.set_param("hostname", "127.0.0.1", params)
        database = self.set_param("database", "kong", params)
        username = self.set_param("username", "kong", params)
        password = self.set_param("password", "kong", params)
        return "host={} dbname={} user={} password={}".format(
            hostname, database, username, password
        )

    def items_str(self, items: List[str]) -> str:
        return ", ".join(items)

    def name_gen(self, quantity: int, entity: str) -> List[List[str]]:
        data = []
        for e in range(quantity):
            data.append([str(uuid.uuid4()), "{}-{}-{}".format(self.prefix, entity, e)])
        return data

    def gen_workspaces(self, quantity) -> List[List[str]]:
        return self.name_gen(quantity, "workspace")

    def get_active_entity_ids(self, entity) -> List[str]:
        active: Dict[str, Any] = self.get_active_entities_data(entity)
        if entity == "workspaces":
            ret: List[str] = [v["id"] for v in active.values()]
        else:
            ret: List[str] = [k for k in active.keys()]
        return ret

    def get_active_workspace_ids(self) -> List[str]:
        entity = "workspaces"
        active: Dict[str, Any] = self.get_active_entities_data(entity)
        ret: List[str] = [v["id"] for v in active.values()]
        return ret

    def string_rep(self, line: str, find: str = "%", replace: str = ",") -> str:
        return re.sub(find, replace, line)

    ################# Create functions ###########################

    def get_active_entities_data(self, entity: str) -> Dict[str, Any]:
        current: Dict[str, Any] = self.db_cache[entity]
        if not current:
            current: Dict[str, Any] = self.get_entities(self.get_items[entity], entity)
            self.db_cache[entity] = current

        active_workspaces: Dict[str, Any] = self.db_cache["workspaces"]
        if not active_workspaces:
            active_workspaces: Dict[str, Any] = self.get_entities(
                self.get_items["workspaces"], "workspaces"
            )
            self.db_cache["workspaces"] = active_workspaces
        required_workspace_names: List[str] = self.required_workspace_names
        if entity == "workspaces":
            active: List[str] = required_workspace_names
        else:
            active: List[str] = [
                v["id"]
                for v in active_workspaces.values()
                if v["name"] in required_workspace_names
            ]

        return self.get_entity_data[entity](current, active, entity)

    def get_active_entity_composit_names(
        self, entity: str, key: str = "composit"
    ) -> List[str]:
        comp: List[str] = self.composit_cache[entity]
        if not comp:
            data: Dict[str, Any] = self.parse_cache[entity]
            if not data:
                data: Dict[str, Any] = self.get_active_entities_data(entity)
                self.parse_cache[entity] = data
            comp: List[str] = [v[key] for v in data.values()]
            self.composit_cache[entity] = comp
        return comp

    def entities_to_create(self, entity: str) -> List[List[str]]:
        key: str = "composit"
        active: List[str] = self.get_active_entity_composit_names(entity, key)
        required: List[str] = self.require_strings[entity]()
        ret = []
        for v in required:
            p = v.split("^")
            name = v
            if name in active:
                continue
            config = "CONFIG"
            if p[0] in self.plugins.keys():
                config = json.dumps(self.plugins[p[0]]["config"])
            ret.append(
                [
                    self.string_rep(self.string_rep(d), "CONFIG", config)
                    for d in self.entity_data_hydrate[entity](v).split(",")
                ]
            )
        return ret

    def workspaces_to_create(self) -> List[List[str]]:
        key: str = "name"
        entity: str = "workspaces"
        active: List[str] = self.get_active_entity_composit_names(entity, key)
        required: List[str] = self.require_strings[entity]()
        ret = []
        for v in required:
            p = v.split("^")
            name = p[1]
            if name in active:
                continue
            ret.append(
                [
                    self.string_rep(d)
                    for d in self.entity_data_hydrate[entity](v).split(",")
                ]
            )
        return ret

    def create_entity(self, entity: str) -> None:
        if entity == "workspaces":
            entities_needed: List[List[str]] = self.workspaces_to_create()
        else:
            entities_needed: List[List[str]] = self.entities_to_create(entity)
        self.insert_into_table(entity, self.insert_items[entity], entities_needed)

    ################ Delete functions ##############################

    def delete_entity(self, entity: str) -> None:
        if entity == "workspaces":
            delete: List[str] = self.get_active_workspace_ids()
        else:
            delete: List[str] = self.get_active_entity_ids(entity)
        self.delete_from_table(entity, delete)

    ################ SQL function ################################

    def get_entities(self, items: List[str], table: str) -> Dict[str, Any]:
        data = {}
        with psycopg.connect(self.db_connect(self.db_params)) as conn:
            with conn.cursor() as cursor:
                with cursor.copy(
                    "COPY (SELECT {} FROM {}) TO STDOUT".format(
                        self.items_str(items), table
                    )
                ) as copy:
                    for row in copy.rows():
                        data[row[0]] = dict(zip(items, row))
        return data

    def insert_into_table(
        self, table: str, items: List[str], data: List[List[str]]
    ) -> None:
        print("{} to create: {}".format(table, len(data)))
        with psycopg.connect(self.db_connect(self.db_params)) as conn:
            with conn.cursor() as cursor:
                with cursor.copy(
                    "COPY {} ({}) FROM STDIN".format(table, self.items_str(items))
                ) as copy:
                    for record in data:
                        copy.write_row(record)
        self.db_cache[table] = self.get_entities(self.get_items[table], table)

    def delete_from_table(
        self, table: str, ids_to_delete: List[str], id_key: str = "id"
    ):
        print("{} to delete: {}".format(table, len(ids_to_delete)))
        with psycopg.connect(self.db_connect(self.db_params)) as conn:
            with conn.cursor() as cursor:
                cursor.execute("CREATE TEMP TABLE to_delete (id uuid);")
                with cursor.copy("COPY to_delete (id) FROM STDIN;") as copy:
                    for id_ in ids_to_delete:
                        copy.write_row((id_,))
                cursor.execute(
                    """\
                    DELETE FROM {} WHERE
                       {} IN (SELECT id FROM to_delete);""".format(
                        table, id_key
                    )
                )
        self.db_cache[table] = self.get_entities(self.get_items[table], table)

    ################ entry functions ##############################

    def create_entities(self) -> None:
        for entity in self.entites:
            self.create_entity(entity)

    def delete_entities(self) -> None:
        for entity in reversed(self.entites):
            self.delete_entity(entity)


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="Kong DB import tool")
    parser.add_argument(
        "--config-file", metavar="path", required=True, help="the path to config file"
    )
    parser.add_argument(
        "--hostname",
        metavar="host",
        required=True,
        default="127.0.0.1",
        help="the hostname for the Kong database",
    )
    parser.add_argument(
        "--database",
        metavar="database",
        required=True,
        default="kong",
        help="the name of the Kong database",
    )
    parser.add_argument(
        "--username",
        metavar="username",
        required=True,
        default="kong",
        help="the username to access the Kong database",
    )
    parser.add_argument(
        "--password",
        metavar="password",
        required=True,
        default="kong",
        help="the password to access the Kong database",
    )
    parser.add_argument(
        "--delete",
        action="store_true",
        help="instead of creating the entities in the kong database delete \
them if this flag is set",
    )
    parser.add_argument(
        "--route-dump",
        action="store_true",
        help="Dump out a JSON list of all route paths (for automated load test)",
    )
    parser.add_argument(
        "--route-dump-location",
        metavar="path",
        required=False,
        help="Dump out a JSON list of all route paths (for automated load test)",
    )

    args = parser.parse_args()
    params = {
        "hostname": args.hostname,
        "database": args.database,
        "username": args.username,
        "password": args.password,
    }
    Runner(
        args.config_file, params, args.delete, args.route_dump, args.route_dump_location
    )
